c/tskit/trees.c

```diff
@@ -7814,75 +7835,455 @@ remove_unextended(edge_list_t **head, edge_list_t **tail)
+        if (self->near_side[e] != self->far_side[e]) {
+            ret = edge_list_append_entry(&self->edges_out_head, &self->edges_out_tail,
+                &self->edge_list_heap, e, 0);
+            if (ret != 0) {
+                ret = TSK_ERR_NO_MEMORY;
+                goto out;
+            }
+        }
+    }
+    for (ex_out = self->edges_out_head; ex_out != NULL; ex_out = ex_out->next) {
+        e = ex_out->edge;
+        self->parent_out[self->edges->child[e]] = self->edges->parent[e];
+        self->next_nodes_edge[self->edges->child[e]] = TSK_NULL;
+        self->next_degree[self->edges->child[e]] -= 1;
+        self->next_degree[self->edges->parent[e]] -= 1;
+    }
+
+    for (tj = tree_pos->in.start; tj != tree_pos->in.stop; tj += self->direction) {
+        e = tree_pos->in.order[tj];
+        // add edge to pending_in
+        ret = edge_list_append_entry(
+            &self->edges_in_head, &self->edges_in_tail, &self->edge_list_heap, e, 0);
+        if (ret != 0) {
+            ret = TSK_ERR_NO_MEMORY;
+            goto out;
+        }
+    }
+    for (ex_in = self->edges_in_head; ex_in != NULL; ex_in = ex_in->next) {
+        e = ex_in->edge;
+        self->parent_in[self->edges->child[e]] = self->edges->parent[e];
+        self->next_nodes_edge[self->edges->child[e]] = e;
+        self->next_degree[self->edges->child[e]] += 1;
+        self->next_degree[self->edges->parent[e]] += 1;
+    }
+
+out:
+    return ret;
+}
+
+static int
+haplotype_extender_add_or_extend_edge(haplotype_extender_t *self, tsk_id_t new_parent,
+    tsk_id_t child, double left, double right)
+{
+    int ret = 0;
+    double there;
+    tsk_id_t old_edge, e_out, old_parent;
+    edge_list_t *ex_in;
+    tsk_id_t e_in;
+
+    there = (self->direction == TSK_DIR_FORWARD) ? right : left;
+    old_edge = self->next_nodes_edge[child];
+    if (old_edge != TSK_NULL) {
+        old_parent = self->edges->parent[old_edge];
     } else {
-        near_side = edges->right;
-        far_side = edges->left;
+        old_parent = TSK_NULL;
+    }
+    if (new_parent != old_parent) {
+        if (self->parent_out[child] == new_parent) {
+            // if our new edge is in edges_out, it should be extended
+            e_out = self->last_nodes_edge[child];
+            self->far_side[e_out] = there;
+            edge_list_set_extended(&self->edges_out_head, e_out);
+        } else {
+            e_out = tsk_edge_table_add_row(
+                self->edges, left, right, new_parent, child, NULL, 0);
+            if (e_out < 0) {
+                ret = (int) e_out;
+                goto out;
+            }
+            /* pointers to left/right might have changed! */
+            if (self->direction == TSK_DIR_FORWARD) {
+                self->near_side = self->edges->left;
+                self->far_side = self->edges->right;
+            } else {
+                self->near_side = self->edges->right;
+                self->far_side = self->edges->left;
+            }
```

Comment:
Maybe allocate the new edge list node here and pass to ``edge_list_append_entry`` rather than passing around the ``edge_list_heap``? It's not much extra code, and will save you some a few unreachable lines. It seemed odd to me on reading to pass the allocator in rather than the actual pointer.